### PR TITLE
Tmsp proof

### PR DESCRIPTION
**Pull request by ethanfrey**
@jaekwon Expose merkle proofs in merkle.Tree and extend tests.

Needed for progress on https://github.com/tendermint/tendermint/issues/355

### CODE REVIEW

File: iavl_proof.go [hunk 1]
```diff
@@ -9,32 +9,42 @@ import (
...
 	}
-	// fmt.Printf("root: %X, computed: %X\n", proof.RootHash, hash)
 	return bytes.Equal(proof.RootHash, hash)
 }
 
+// LoadProof will deserialize a IAVLProof from bytes
+func LoadProof(data []byte) (*IAVLProof, error) {
+	// TODO: make go-wire never panic
+	n, err := int(0), error(nil)
+	proof := wire.ReadBinary(&IAVLProof{}, bytes.NewBuffer(data), proofLimit, &n, &err).(*IAVLProof)
```
Comment (akalyaev) [inline iavl_proof.go]: And what if a proof will be bigger than `proofLimit`? Or it cannot happen?

Comment (ethanfrey) [inline iavl_proof.go]: 64 KB should be plenty.  I got plenty of panics deserializing random bytes when it was larger.  (tests run MutateByteSlice and then this invalid info crashes go-wire unless you give it some sort of sanity check).

We could raise it, maybe 1MB?  Just make sure it doesn't overflow the heap...

Comment (jaekwon) [inline iavl_proof.go]: 64KB seems fine here.

**Comment by jaekwon** [comment]
Left a question but otherwise LGTM.

**Pull request merged**